zerver/lib/streams.py

```diff
@@ -55,7 +55,7 @@ def access_stream_common(user_profile: UserProfile, stream: Stream,
         sub = None
 
     # If the stream is in your realm and public, you can access it.
-    if stream.is_public():
+    if stream.is_public() and not user_profile.is_guest:
```

Comment:
The tests you have so far just cover this code path; we should test the others as well.